File: 0981-time-based-key-value-store/0981-time-based-key-value-store.java
class TimeMap {
    Map<String, List<Pair>> map;

    public TimeMap() {
        map = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        map.putIfAbsent(key, new ArrayList<>());
        List<Pair> pairs = map.get(key);
        pairs.add(new Pair(timestamp, value));
    }
    
    public String get(String key, int timestamp) {
        if (!map.containsKey(key)) return "";

        List<Pair> pairs = map.get(key);
        int left = 0;
        int right = pairs.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (pairs.get(mid).timestamp > timestamp) {
                right = mid - 1;
            } else if (pairs.get(mid).timestamp < timestamp) {
                left = mid + 1;
            } else {
                return pairs.get(mid).value;
            }
        }

        if (right < 0) return "";

        return pairs.get(right).value;
    }

    private class Pair {
        int timestamp;
        String value;

        public Pair(int timestamp, String value) {
            this.timestamp = timestamp;
            this.value = value;
        }
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.kt
class TimeMap() {
    private val map = HashMap<String, MutableList<Pair<Int, String>>>()

    fun set(key: String, value: String, timestamp: Int) {
        map.getOrPut(key) { mutableListOf() }.add(timestamp to value)
    }

    fun get(key: String, timestamp: Int): String {
        val pairs = map[key] ?: return ""
        var left = 0
        var right = pairs.size - 1

        while (left <= right) {
            val mid = left + (right - left) / 2

            when {
                pairs[mid].first > timestamp -> right = mid - 1
                pairs[mid].first < timestamp -> left = mid + 1
                else -> return pairs[mid].second
            }
        }

        return if (right < 0) "" else pairs[right].second
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * var obj = TimeMap()
 * obj.set(key,value,timestamp)
 * var param_2 = obj.get(key,timestamp)
 */